libp2p/security/noise/patterns.py

```diff
@@ -0,0 +1,116 @@
+from abc import ABC, abstractmethod
+
+from noise.connection import Keypair as NoiseKeypair
+from noise.connection import NoiseConnection as NoiseState
+
+from libp2p.crypto.keys import PrivateKey
+from libp2p.network.connection.raw_connection_interface import IRawConnection
+from libp2p.peer.id import ID
+from libp2p.security.secure_conn_interface import ISecureConn
+
+from .connection import NoiseConnection
+
+# FIXME: Choose a serious bound number.
+NUM_BYTES_TO_READ = 2048
+
+
+# TODO: Merged into `BasePattern`?
+class PreHandshakeConnection:
+    conn: IRawConnection
+
+    def __init__(self, conn: IRawConnection) -> None:
+        self.conn = conn
+
+    async def write_msg(self, data: bytes) -> None:
+        # TODO:
+        await self.conn.write(data)
+
+    async def read_msg(self) -> bytes:
+        return await self.conn.read(NUM_BYTES_TO_READ)
+
+
+class IPattern(ABC):
+    @abstractmethod
+    async def handshake_inbound(self, conn: IRawConnection) -> ISecureConn:
+        ...
+
+    @abstractmethod
+    async def handshake_outbound(
+        self, conn: IRawConnection, remote_peer: ID
+    ) -> ISecureConn:
+        ...
+
+
+class BasePattern(IPattern):
+    protocol_name: bytes
+    noise_static_key: PrivateKey
+    local_peer: ID
+    libp2p_privkey: PrivateKey
+
+    def create_noise_state(self) -> NoiseState:
+        noise_state = NoiseState.from_name(self.protocol_name)
+        noise_state.set_keypair_from_private_bytes(
+            NoiseKeypair.STATIC, self.noise_static_key.to_bytes()
+        )
+        return noise_state
+
+
+class PatternXX(BasePattern):
+    def __init__(
+        self, local_peer: ID, libp2p_privkey: PrivateKey, noise_static_key: PrivateKey
+    ) -> None:
+        self.protocol_name = b"Noise_XX_25519_ChaChaPoly_SHA256"
```

Comment:
is this the specific noise suite we are using in `libp2p-noise` according to that spec?

Reply:
Yes, it's the protocol we use in the XX pattern. Ref: https://github.com/libp2p/specs/tree/master/noise#valid-noise-protocol-names